sycl/plugins/level_zero/pi_level_zero.hpp

```diff
@@ -731,6 +749,41 @@ struct _pi_queue : _pi_object {
   // asked to not transfer the ownership to SYCL RT.
   bool OwnZeCommandQueue;
 
+  // Indicates if we create an event for each command or use eventless mode for
+  // the queue. The eventless mode is used for SYCL in-order queue with
+  // discard_events property. Since we don't create events for some commands we
+  // guarantee in-order semantics inside command-list by using barriers and
+  // create a special event for the last barrier of the command-list to
+  // guarantee the semantics between command-lists.
+  bool EventlessMode = false;
+
+  // Indicates if the previous command was submitted into the copy engine. it is
+  // used in eventless mode to identify that we changed compute queue to copy
+  // queue or vice versa. if the change happened then we create an event to
+  // maintain the order between the two queues.
+  bool IsPrevCopyEngine = false;
+
+  // It helps to skip creating the event and the last barrier for the
+  // command-list in eventless mode if we do QueueFinish or piQueueRelease
+  bool SkipLastEventInEventlessMode = false;
+
+  // Keeps a number of commands submitted into last command-list that don't have
+  // events. It can be negative which means that we are in eventless mode and
+  // the last command has changed copy engine to compute or vice versa and this
+  // command produces event.
+  int NumEventlessCmdsInLastCmdList{0};
+
+  // Keeps last special event created for internal purpose to ensure in-order
+  // semantics between command-lists. The plugin is a holder of this event and
+  // SYCL RT knows nothing about it. This special event will be destroyed and
+  // returned back in pool by EventRelease in resetCommandList. this is used to
+  // compare against LastCommandEvent to ensure that event-dependency will only
+  // be added between command-lists.
+  pi_event LastEventInPrevCmdList = nullptr;
+
+  // Keeps command-list of the previous command to use in EventlessMode.
+  pi_command_list_ptr_t LastCommandList{};
+
```

Comment:
I feel that this is adds massive redundant state to the PI layer. Please outline the approach are are trying to implement in the description of this PR

Reply:
Done.